### PR TITLE
Fix/reshape combine

### PR DESCRIPTION
**Pull request by xhuohai**
combine reshape with other op

### CODE REVIEW

File: src/Nncase.Passes/Rules/Neutral/CombineReshape.cs [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+    }
+
+    /// <inheritdoc/>
+    public IPattern Pattern { get; init; }
+
+    private Expr? GetReplace(Binary binary, Expr x, Expr y, Expr shape)
+    {
+        if (x.CheckedShape == y.CheckedShape)
+        {
+            return Reshape(Binary(binary.BinaryOp, x, y), shape);
```
Comment (aaltonenzhang) [inline src/Nncase.Passes/Rules/Neutral/CombineReshape.cs:51]: please create Binary using Binary(binary.BinaryOp, x, y).InheritMetaData(binaryCall)

File: src/Nncase.Passes/Rules/Neutral/CombineTranspose.cs [hunk 2]
```diff
@@ -89,6 +94,11 @@ public CombineConstBinaryTranspose()
 
         if (y is Const)
         {
+            if (y.CheckedShape.Rank == 0)
+            {
+                return Transpose(Binary(binary.BinaryOp, x, y), perm);
```
Comment (aaltonenzhang) [inline src/Nncase.Passes/Rules/Neutral/CombineTranspose.cs:99]: please create Binary using Binary(binary.BinaryOp, x, y).InheritMetaData(binaryCall)

File: src/Nncase.Passes/Rules/Neutral/CombineTranspose.cs [hunk 3]
```diff
@@ -73,6 +73,11 @@ public CombineConstBinaryTranspose()
 
         if (x is Const)
         {
+            if (x.CheckedShape.Rank == 0)
+            {
+                return Transpose(Binary(binary.BinaryOp, x, y), perm);
```
Comment (aaltonenzhang) [inline src/Nncase.Passes/Rules/Neutral/CombineTranspose.cs:78]: please create Binary using Binary(binary.BinaryOp, x, y).InheritMetaData(binaryCall)

File: src/Nncase.Passes/Rules/Neutral/CombineTranspose.cs (hunk 2 above)
Comment (aaltonenzhang) [inline src/Nncase.Passes/Rules/Neutral/CombineTranspose.cs:99]: please create Binary using Binary(binary.BinaryOp, x, y).InheritMetaData(binaryCall)

File: src/Nncase.Passes/Rules/Neutral/CombineTranspose.cs (hunk 3 above)
Comment (aaltonenzhang) [inline src/Nncase.Passes/Rules/Neutral/CombineTranspose.cs:78]: please create Binary using Binary(binary.BinaryOp, x, y).InheritMetaData(binaryCall)

File: src/Nncase.Passes/Rules/Neutral/CombineReshape.cs [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+    public IPattern Pattern { get; init; }
+
+    private Expr? GetReplace(Binary binary, Call call, IReadOnlyList<Expr> callParams, Expr input, TensorConst constInput, TensorConst shape)
+    {
+        var oldShape = shape.Value.ToArray<int>();
+        var significantShape = oldShape.Where(x => x > 1).ToArray();
+
+        if (constInput.CheckedShape.Rank == 0)
+        {
+            var res = Reshape(Binary(binary.BinaryOp, input, constInput), shape);
```
Comment (aaltonenzhang) [inline src/Nncase.Passes/Rules/Neutral/CombineReshape.cs:86]: please create Binary using Binary(binary.BinaryOp, x, y).InheritMetaData(binaryCall)

File: src/Nncase.Passes/Rules/Neutral/CombineReshape.cs [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+        var significantInputShape = input.CheckedShape.ToValueArray().Where(x => x > 1).ToArray();
+        bool leftConst = ReferenceEquals(callParams[0], constInput);
+        var constSize = constInput.CheckedShape.ToValueArray()[0];
+        if (significantShape.SequenceEqual(significantInputShape) && oldShape[^1] == constSize)
+        {
+            var broadcastIndex = Array.LastIndexOf(input.CheckedShape.ToValueArray(), constSize);
+            var newConstShape = Enumerable.Repeat(1, input.CheckedShape.Rank - 1 - broadcastIndex).ToList();
+            newConstShape.Insert(0, constSize);
+
+            var res = Reshape(Binary(binary.BinaryOp, leftConst ? Reshape(constInput, newConstShape.ToArray()) : input, leftConst ? input : Reshape(constInput, newConstShape.ToArray())), call.CheckedShape);
```
Comment (aaltonenzhang) [inline src/Nncase.Passes/Rules/Neutral/CombineReshape.cs:100]: the save

File: src/Nncase.Passes/Rules/Neutral/CombineReshape.cs [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+        IsUnary(
+            "unary",
+            "call",
+            _ => true,
+            IsReshape(IsWildcard("input"), IsWildcard("shape")));
+
+    private Expr? GetReplace(Unary unary, Expr input, Expr shape)
+    {
+        return Reshape(
+            Unary(unary.UnaryOp, input),
```
Comment (aaltonenzhang) [inline src/Nncase.Passes/Rules/Neutral/CombineReshape.cs:126]: the same

File: src/Nncase.Passes/Rules/Neutral/CombineReshape.cs [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+    private Expr? GetReplace(ActivationOp activation, Expr input, IReadOnlyList<Expr> parameters, Expr shape)
+    {
+        // TODO: Not support PRelu for now.
+        if (activation is PRelu)
+        {
+            return null;
+        }
+
+        return Reshape(
+            new Call(activation, new Expr[] { input }.Concat(parameters.Skip(1)).ToArray()),
```
Comment (aaltonenzhang) [inline src/Nncase.Passes/Rules/Neutral/CombineReshape.cs:160]: the same, new activation need name.

File: src/Nncase.Passes/Rules/Neutral/CombineTranspose.cs (hunk 3 above)
Comment (xhuohai) [reply src/Nncase.Passes/Rules/Neutral/CombineTranspose.cs:79]: changed

**Review by aaltonenzhang — APPROVED** [review]

**Review by zhen8838 — APPROVED** [review]

**Pull request merged**